prow/gerrit/reporter/reporter.go

```diff
@@ -62,33 +65,83 @@ func (c *Client) ShouldReport(pj *v1.ProwJob) bool {
 	}
 
 	// has gerrit metadata (scheduled by gerrit adapter)
-	return pj.ObjectMeta.Annotations["gerrit-id"] != "" &&
-		pj.ObjectMeta.Annotations["gerrit-instance"] != "" &&
-		pj.ObjectMeta.Labels["gerrit-revision"] != ""
+	if pj.ObjectMeta.Annotations[gerrit.GerritID] == "" ||
+		pj.ObjectMeta.Annotations[gerrit.GerritInstance] == "" ||
+		pj.ObjectMeta.Labels[gerrit.GerritRevision] == "" {
+		return false
+	}
+
+	// has other jobs on the same revision finished
+	selector := labels.Set{gerrit.GerritRevision: pj.ObjectMeta.Labels[gerrit.GerritRevision]}
+	pjs, err := c.lister.List(selector.AsSelector())
+	if err != nil {
+		logrus.WithError(err).Errorf("Cannot list prowjob with selector %v", selector)
+		return false
+	}
+
+	for _, pj := range pjs {
+		if pj.Status.State == v1.TriggeredState || pj.Status.State == v1.PendingState {
+			// not done yet
+			return false
+		}
+	}
+
+	return true
 }
 
 // Report will send the current prowjob status as a gerrit review
 func (c *Client) Report(pj *v1.ProwJob) error {
-	// TODO(krzyzacy): we should also do an aggregate report, as golang does in their repo
-	// see https://go-review.googlesource.com/c/go/+/132155
-	// also add ability to set code-review labels
-	// ref: https://github.com/kubernetes/test-infra/issues/9433
+	// If you are hitting here, which means the entire patchset has been finished :-)
 
-	if pj.Spec.Refs == nil {
-		return errors.New("no pj.Spec.Refs, not a presubmit job (should not happen?!)")
+	// list all prowjobs in the patchset
+	selector := labels.Set{gerrit.GerritRevision: pj.ObjectMeta.Labels[gerrit.GerritRevision]}
+	pjsOnRevision, err := c.lister.List(selector.AsSelector())
+	if err != nil {
+		logrus.WithError(err).Errorf("Cannot list prowjob with selector %v", selector)
+		return err
 	}
 
-	message := fmt.Sprintf("Job %s finished with %s\n Gubernator URL: %s", pj.Spec.Job, pj.Status.State, pj.Status.URL)
+	// generate an aggregated report:
+	total := len(pjsOnRevision)
+	success := 0
+	message := ""
+
+	for _, pjOnRevision := range pjsOnRevision {
+		if pjOnRevision.Status.PrevReportStates[c.GetName()] == pjOnRevision.Status.State {
+			logrus.Infof("Revision %s has been reported already", pj.ObjectMeta.Labels[gerrit.GerritRevision])
+			return nil
+		}
+
+		if pjOnRevision.Status.State == v1.SuccessState {
+			success++
+		}
+
+		message = fmt.Sprintf("%s\nJob %s finished with %s -- URL: %s", message, pjOnRevision.Spec.Job, pjOnRevision.Status.State, pjOnRevision.Status.URL)
+	}
+
+	message = fmt.Sprintf("%d out of %d jobs passed!\n%s", success, total, message)
```

Comment:
nit: can we kill the exclamation point?

also, wouldn't it be more helpful to report that 7 out of 34 jobs failed rather than 27 out of 34 jobs passed? Aka people mostly on care about failures.

Reply:
:joy: I feel more excited when I see `34 out of 34 jobs passed`? I'm fine with either way.

Reply:
good point, positivity is good